File: packages/shared-kmp/features/auth/src/commonMain/kotlin/dev/forcetower/melon/feature/auth/domain/repository/AuthRepository.kt
package dev.forcetower.melon.feature.auth.domain.repository

import dev.forcetower.melon.core.common.Outcome
import dev.forcetower.melon.core.session.domain.model.User
import dev.forcetower.melon.feature.auth.domain.model.LoginError
import dev.forcetower.melon.feature.auth.domain.model.PasskeyAssertion
import dev.forcetower.melon.feature.auth.domain.model.PasskeyChallenge

internal interface AuthRepository {
    suspend fun login(username: String, password: String): Outcome<User, LoginError>

    suspend fun beginPasskeyLogin(username: String?): Outcome<PasskeyChallenge, LoginError>

    suspend fun completePasskeyLogin(
        sessionId: String,
        assertion: PasskeyAssertion,
    ): Outcome<User, LoginError>
}

File: apps/ios/UNESKit/Sources/UNESKit/Domain/Repositories/AuthRepository.swift
import ComposableArchitecture

/// Login against apps/api with SAGRES credentials or a passkey.
/// Successful logins persist the session in the `SessionStore`.
@DependencyClient
struct AuthRepository: Sendable {
    var login: @Sendable (_ username: String, _ password: String) async throws -> Session
    var beginPasskeyLogin: @Sendable () async throws -> PasskeyChallenge
    var completePasskeyLogin: @Sendable (_ sessionId: String, _ assertion: PasskeyAssertion) async throws -> Session
}

extension AuthRepository: TestDependencyKey {
    static let testValue = AuthRepository()

    static let previewValue = AuthRepository(
        login: { _, _ in .preview },
        beginPasskeyLogin: {
            PasskeyChallenge(sessionId: "preview", challenge: "", rpId: "melon.forcetower.dev", allowedCredentialIds: [])
        },
        completePasskeyLogin: { _, _ in .preview }
    )
}

extension DependencyValues {
    var authRepository: AuthRepository {
        get { self[AuthRepository.self] }
        set { self[AuthRepository.self] = newValue }
    }
}
